meta_client/src/types.rs

```diff
@@ -162,18 +162,58 @@ pub struct NodeInfo {
     pub shard_infos: Vec<ShardInfo>,
 }
 
+/// The status changes of a shard as following:
+///
+///```plaintext
+///   ┌────┐
+///   │Init│
+///   └──┬─┘
+///   ___▽___
+///  ╱       ╲     ┌─────┐
+/// ╱ Opening ╲____│Ready│
+/// ╲         ╱yes └──┬──┘
+///  ╲_______╱    ┌───▽──┐
+///               │Frozen│
+///               └──────┘
+/// ```
+/// When an open request comes in, shard can only be opened when it's in
+/// - `Init`, which means it has not been opened before.
+/// - `Opening`, which means it has been opened before, but failed.
+#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize)]
+pub enum ShardStatus {
+    /// Created, but not opened
+    #[default]
+    Init,
+    /// In opening
+    Opening,
+    /// Healthy
+    Ready,
+    /// Further updates are prohibited
+    Frozen,
+}
+
 #[derive(Debug, Default, Clone, PartialEq, Eq, Serialize)]
 pub struct ShardInfo {
```

Comment:
I guess it may better to move it back to `cluster`, and we define a struct just for sending in `meta_client`.